vm/src/builtins/type.rs

```diff
@@ -523,6 +523,57 @@ impl PyType {
         Ok(())
     }
 
+    #[pygetset(magic)]
+    fn annotations(&self, vm: &VirtualMachine) -> PyResult<PyObjectRef> {
+        if !self.slots.flags.has_feature(PyTypeFlags::HEAPTYPE) {
+            return Err(vm.new_attribute_error(format!(
+                "type object '{}' has no attribute '__annotations__'",
+                self.name()
+            )));
+        }
+
+        Ok(self
+            .attributes
+            .read()
+            .get(identifier!(vm, __annotations__))
+            .cloned()
+            .unwrap_or_else(|| {
+                self.attributes
+                    .write()
+                    .insert(identifier!(vm, __annotations__), vm.ctx.new_dict().into())
+                    .unwrap_or_else(|| vm.ctx.new_dict().into())
+            }))
+    }
+
+    #[pygetset(magic, setter)]
+    fn set_annotations(&self, value: Option<PyObjectRef>, vm: &VirtualMachine) -> PyResult<()> {
+        if !self.slots.flags.has_feature(PyTypeFlags::IMMUTABLETYPE) {
+            return Err(vm.new_type_error(format!(
+                "cannot set '__annotations__' attribute of immutable type '{}'",
+                self.name()
+            )));
+        }
+
+        let __annotations__ = identifier!(vm, __annotations__);
+        if let Some(value) = value {
+            self.attributes.write().insert(__annotations__, value);
+        } else {
+            self.attributes
+                .read()
+                .get(__annotations__)
+                .cloned()
+                .ok_or_else(|| {
+                    vm.new_attribute_error(format!(
+                        "'{}' object has no attribute '__annotations__'",
+                        self.name()
+                    ))
+                })?;
+            self.attributes.write().remove(__annotations__);
```

Comment:
maybe a single write lock? double-deleting `__annotations__` will not be a common case.